File: app/src/main/java/com/egecius/rxjava2_demo_2/rx/maybe/MaybeExamples.java
package com.egecius.rxjava2_demo_2.rx.maybe;

import android.support.annotation.Nullable;

import com.egecius.rxjava2_demo_2.rx.EgisException;

import java.util.List;
import java.util.NoSuchElementException;

import io.reactivex.Completable;
import io.reactivex.CompletableSource;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.subjects.BehaviorSubject;

public class MaybeExamples {

    public boolean isFlatMapCompletableExecuted;
    public boolean isAndThenExecuted;

    public Single<Integer> maybeToSingle(Integer param) {

        return Maybe.just(param)
                .filter(integer -> integer < 5)
                .toSingle();
    }

    /**
     * defaultIfEmpty() surprisingly returns Empty, even though it could return Single, since it
     * won't be empty now.
     */
    public Single<Integer> defaultIfEmpty(@Nullable Integer param) {

        Maybe<Integer> maybe;

        if (param == null) {
            maybe = Maybe.empty();
        } else {
            maybe = Maybe.just(param);
        }

        return maybe
                .defaultIfEmpty(-1)
                .toSingle();
    }

    /** flatMapCompletable will not get executed if origin Maybe is empty */
    public Completable maybeFlatMapCompletable(Iterable<Integer> integers) {
        return Observable.fromIterable(integers)
                .firstElement()
                .flatMapCompletable(integer -> getCompletable());
    }

    private CompletableSource getCompletable() {
        isFlatMapCompletableExecuted = true;
        return Completable.complete();
    }

    /** andThen() gets executed even if Function in flatMapCompletable() before it did not get
     * executed  */
    public Completable maybeFlatMapCompletableAndThen(Iterable<Integer> integers) {
        return Observable.fromIterable(integers)
                .firstElement()
                .flatMapCompletable(integer -> getCompletable())
                .andThen(myAndThen());
    }

    private CompletableSource myAndThen() {
        isAndThenExecuted = true;
        return Completable.complete();
    }

    public Maybe<String> filterEvenAndMapToString(List<Integer> list) {

        return Observable.fromIterable(list)
                .firstElement()
                .filter(this::isEven)
                .map(String::valueOf);
    }

    private boolean isEven(Integer integer) {
        return integer % 2 == 0;
    }

    /** flatMapSingle fails with {@link NoSuchElementException} when it receives no value, i.e.
     * when Maybe completes rather than calls onSuccess() */
    public Single<String> flatMapSingle(List<Integer> list) {

        return Observable.fromIterable(list)
                .firstElement()
                .flatMapSingle(integer -> Single.just(String.valueOf(integer)));
    }

    Completable maybeIgnoreElementFromSuccess() {
        return Maybe.just(5).ignoreElement();
    }

    Completable maybeIgnoreElementFromEmpty() {
        return Maybe.empty().ignoreElement();
    }

    Completable maybeIgnoreElementFromError() {
        return Maybe.error(new EgisException())
                .ignoreElement();
    }

    Completable flatMapCompletable() {
        BehaviorSubject<Boolean> behaviorSubject = getBehaviorSubjectWithDefault();
        return behaviorSubject.flatMapCompletable(aBoolean -> Completable.complete());
    }

    Completable flatMapCompletableWithEmission() {
        BehaviorSubject<Boolean> behaviorSubject = getBehaviorSubjectWithDefault();
        behaviorSubject.onNext(true);
        Completable flatMapCompletable = behaviorSubject.flatMapCompletable(
                aBoolean -> Completable.complete());

        behaviorSubject.onNext(true);

        return flatMapCompletable;
    }

    BehaviorSubject<Boolean> getBehaviorSubjectWithDefault() {
        return BehaviorSubject.createDefault(true);
    }

    Completable firstElementIgnoreElement() {
        BehaviorSubject<Boolean> behaviorSubject = getBehaviorSubjectWithDefault();
        return behaviorSubject
                .firstElement()
                .ignoreElement();
    }
}

File: app/src/main/java/com/egecius/rxjava2_demo_2/rx/maybe/MaybeExamples.kt
package com.egecius.rxjava2_demo_2.rx.maybe

import com.egecius.rxjava2_demo_2.rx.EgisException
import io.reactivex.*
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import java.util.*

class MaybeExamples {

    var isFlatMapCompletableExecuted: Boolean = false
    var isAndThenExecuted: Boolean = false

    private val completable: CompletableSource
        get() {
            isFlatMapCompletableExecuted = true
            return Completable.complete()
        }

    val behaviorSubjectWithDefault: BehaviorSubject<Boolean>
        get() = BehaviorSubject.createDefault(true)

    fun maybeToSingle(param: Int?): Single<Int> {

        return Maybe.just(param!!)
                .filter { integer -> integer < 5 }
                .toSingle()
    }

    /**
     * defaultIfEmpty() surprisingly returns Empty, even though it could return Single, since it
     * won't be empty now.
     */
    fun defaultIfEmpty(param: Int?): Single<Int> {

        val maybe: Maybe<Int>

        if (param == null) {
            maybe = Maybe.empty()
        } else {
            maybe = Maybe.just(param)
        }

        return maybe
                .defaultIfEmpty(-1)
                .toSingle()
    }

    /** flatMapCompletable will not get executed if origin Maybe is empty  */
    fun maybeFlatMapCompletable(integers: Iterable<Int>): Completable {
        return Observable.fromIterable(integers)
                .firstElement()
                .flatMapCompletable { integer -> completable }
    }

    /** andThen() gets executed even if Function in flatMapCompletable() before it did not get
     * executed   */
    fun maybeFlatMapCompletableAndThen(integers: Iterable<Int>): Completable {
        return Observable.fromIterable(integers)
                .firstElement()
                .flatMapCompletable { integer -> completable }
                .andThen(myAndThen())
    }

    private fun myAndThen(): CompletableSource {
        isAndThenExecuted = true
        return Completable.complete()
    }

    fun filterEvenAndMapToString(list: List<Int>): Maybe<String> {

        return Observable.fromIterable(list)
                .firstElement()
                .filter({ this.isEven(it) })
                .map({ it.toString() })
    }

    private fun isEven(integer: Int?): Boolean {
        return integer!! % 2 == 0
    }

    /** flatMapSingle fails with [NoSuchElementException] when it receives no value, i.e.
     * when Maybe completes rather than calls onSuccess()  */
    fun flatMapSingle(list: List<Int>): Single<String> {

        return Observable.fromIterable(list)
                .firstElement()
                .flatMapSingle { integer -> Single.just(integer.toString()) }
    }

    fun maybeIgnoreElementFromSuccess(): Completable {
        return Maybe.just(5).ignoreElement()
    }

    fun maybeIgnoreElementFromEmpty(): Completable {
        return Maybe.empty<Any>().ignoreElement()
    }

    fun maybeIgnoreElementFromError(): Completable {
        return Maybe.error<Any>(EgisException())
                .ignoreElement()
    }

    fun flatMapCompletable(): Completable {
        val behaviorSubject = behaviorSubjectWithDefault
        return behaviorSubject.flatMapCompletable { aBoolean -> Completable.complete() }
    }

    fun flatMapCompletableWithEmission(): Completable {
        val behaviorSubject = behaviorSubjectWithDefault
        behaviorSubject.onNext(true)
        val flatMapCompletable = behaviorSubject.flatMapCompletable { aBoolean -> Completable.complete() }

        behaviorSubject.onNext(true)

        return flatMapCompletable
    }

    fun firstElementIgnoreElement(): Completable {
        val behaviorSubject = behaviorSubjectWithDefault
        return behaviorSubject
                .firstElement()
                .ignoreElement()
    }
}
